src/core/geometry/qgsgeometry.cpp

```diff
@@ -538,6 +538,91 @@ bool QgsGeometry::deleteVertex( int atVertex )
   return d->geometry->deleteVertex( id );
 }
 
+bool QgsGeometry::convertVertex( int atVertex )
+{
+
+  if ( !d->geometry )
+    return false;
+
+  QgsVertexId id;
+  if ( !vertexIdFromVertexNr( atVertex, id ) )
+    return false;
+
+  detach();
+
+  QgsAbstractGeometry *geom = d->geometry.get();
+
+  // If the geom is a collection, we get the concerned part, otherwise, the part is just the whole geom
+  QgsAbstractGeometry *part = nullptr;
+  QgsGeometryCollection *owningCollection = dynamic_cast<QgsGeometryCollection *>( geom );
+  if ( owningCollection != nullptr )
+    part = owningCollection->geometryN( id.part );
+  else
+    part = geom;
+
+  // If the part is a polygon, we get the concerned ring, otherwise, the ring is just the whole part
+  QgsAbstractGeometry *ring = nullptr;
+  QgsCurvePolygon *owningPolygon = dynamic_cast<QgsCurvePolygon *>( part );
+  if ( owningPolygon != nullptr )
+    ring = ( id.ring == 0 ) ? owningPolygon->exteriorRing() : owningPolygon->interiorRing( id.ring - 1 );
+  else
+    ring = part;
+
+  // If the ring is not a curve, we're probably on a point geometry
+  QgsCurve *curve = dynamic_cast<QgsCurve *>( ring );  // TODO dynamic_cast -> geom_cast
+  if ( curve == nullptr )
+    return false;
+
+  bool success = false;
+  QgsCompoundCurve *cpdCurve  = dynamic_cast<QgsCompoundCurve *>( curve );
+  if ( cpdCurve != nullptr )
+  {
+    // If the geom is a already compound curve, we convert inplace, and we're done
+    success = cpdCurve->convertVertex( id );
+  }
+  else
+  {
+    // TODO : move this block before the above, so we call convertVertex only in one place
+    // If the geom is a linestring or cirularstring, we create a compound curve
+    QgsCompoundCurve *cpdCurve = new QgsCompoundCurve();
```

Comment:
```suggestion
    std::unique_ptr<QgsCompoundCurve> cpdCurve = std::make_unique<QgsCompoundCurve>();
```
Better safe than sorry, this will make sure nothing is leaked.
And avoid the extra copy on line 598 with the "review plz" comment